src/components/db-explorer.tsx

```diff
@@ -0,0 +1,249 @@
+import { useDebounce } from "use-debounce";
+import Bug from "../bug.svg";
+import { useDataFile } from "../hooks";
+import { ErrorState } from "./error-state";
+import { LoadingState } from "./loading-state";
+import { Spinner } from "./spinner";
+
+interface Props {
+  sha: string;
+  filename: string;
+  owner: string;
+  name: string;
+}
+
+interface DBExplorerInnerProps {
+  content: string;
+  filename: string;
+  extension: string;
+  sha: string;
+}
+
+const VALID_EXTENSIONS = ["csv", "json"];
+
+function ErrorFallback(props: FallbackProps) {
+  const { error, resetErrorBoundary } = props;
+  return (
+    <ErrorState img={Bug} alt={error.message}>
+      <p>{error?.message}</p>
+      <div className="mt-4">
+        <button
+          className="inline-flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-gray-900 hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-gray-500"
+          onClick={resetErrorBoundary}
+        >
+          Reset Query
+        </button>
+      </div>
+    </ErrorState>
+  );
+}
+
+function DBExplorerInner(props: DBExplorerInnerProps) {
+  const { content, extension, filename, sha } = props;
+  const filenameWithoutExtension = filename.split(".").slice(0, -1).join(".");
+  const connectionRef = useRef<duckdb.AsyncDuckDBConnection | null>(null);
+  const [query, setQuery] = useState("");
+  const [debouncedQuery] = useDebounce(query, 500);
+  const [dbStatus, setDbStatus] = useState<"error" | "idle" | "success">(
+    "idle"
+  );
+
+  const execQuery = async (query: string) => {
+    if (!connectionRef.current) return;
+    const queryRes = await connectionRef.current.query(query);
+    const asArray = queryRes.toArray();
+
+    return {
+      numRows: queryRes.numRows,
+      numCols: queryRes.numCols,
+      results: asArray.map((row) => {
+        return row.toJSON();
+      }),
+    };
+  };
+
+  const { data, status, error } = useQuery(
+    ["query-results", filename, sha, debouncedQuery],
+    () => execQuery(debouncedQuery),
+    {
+      refetchOnWindowFocus: false,
+      retry: false,
+      enabled: dbStatus === "success",
+    }
+  );
+
+  useEffect(() => {
+    const initDuckDb = async () => {
+      const MANUAL_BUNDLES: duckdb.DuckDBBundles = {
+        mvp: {
+          mainModule: duckdb_wasm,
+          mainWorker: eh_worker,
+        },
+        eh: {
+          mainModule: duckdb_wasm_next,
+          mainWorker: eh_worker,
+        },
+      };
+      const bundle = await duckdb.selectBundle(MANUAL_BUNDLES);
+      const worker = new Worker(bundle.mainWorker!);
+      const logger = new duckdb.ConsoleLogger();
+      const db = new duckdb.AsyncDuckDB(logger, worker);
+      await db.instantiate(bundle.mainModule, bundle.pthreadWorker);
+
+      const c = await db.connect();
+      connectionRef.current = c;
+
+      try {
+        await db.registerFileText(filename, content);
+        extension === "csv"
```

Comment:
I'd add a switch statement since you may want to support other file type (e.g. parquet) in the future.